runtime/src/accounts_db.rs

```diff
@@ -7752,11 +7819,19 @@ impl AccountsDb {
                         .iter()
                         .skip(skip)
                         .take(UNREF_ACCOUNTS_BATCH_SIZE)
-                        .map(|(_slot, pubkey)| pubkey),
-                    |_pubkey, _slots_refs| /* unused */AccountsIndexScanResult::Unref,
+                        .filter_map(|(_slot, pubkey)| {
+                            // filter out pubkeys that have already been removed from the accounts index in a previous step
+                            let already_removed =
+                                pubkeys_removed_from_accounts_index.contains(pubkey);
+                            (!already_removed).then_some(pubkey)
```

Comment:
This is the place that changed behavior. If the pubkey should not be unref'd, then it gets filtered out here BEFORE we unref it. Previously, we always unref'd any key that was in `purged_slot_pubkeys`. If we had already removed that pubkey from the accounts index, then tx processing could have re-added the same pubkey. We would then mess up the refcount.